File: android-sdk/sdk/src/main/java/com/adlcom/mxl/sdk/network/NetworkMonitor.kt
package com.adlcom.mxl.sdk.network

import com.adlcom.mxl.sdk.config.SdkConfiguration
import com.adlcom.mxl.sdk.internal.SdkState
import com.adlcom.mxl.sdk.internal.getSdkState
import com.adlcom.mxl.sdk.logging.Logger
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import java.io.IOException

/**
 * Network monitor interceptor for OkHttp.
 */
class NetworkMonitorInterceptor(
    private val configuration: SdkConfiguration
) : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        if (!configuration.enableNetworkMonitoring) {
            return chain.proceed(chain.request())
        }

        val request = chain.request()
        val requestStartTime = System.currentTimeMillis()

        var response: Response? = null
        var exception: Exception? = null

        try {
            response = chain.proceed(request)
            return response
        } catch (e: IOException) {
            exception = e
            throw e
        } finally {
            val requestEndTime = System.currentTimeMillis()
            val duration = requestEndTime - requestStartTime

            recordNetworkEvent(
                request = request,
                response = response,
                exception = exception,
                duration = duration
            )
        }
    }

    /**
     * Record network event.
     */
    private fun recordNetworkEvent(
        request: Request,
        response: Response?,
        exception: Exception?,
        duration: Long
    ) {
        val sdkState = getSdkState() ?: return

        val requestBodySize = request.body?.contentLength() ?: 0L
        val responseBodySize = response?.body?.contentLength() ?: 0L
        val statusCode = response?.code ?: -1
        val isSuccess = response?.isSuccessful ?: false

        val networkEvent = NetworkEvent(
            sessionId = sdkState.sessionManager.getSessionId() ?: "unknown",
            userId = sdkState.sessionManager.getUserId(),
            url = request.url.toString(),
            method = request.method,
            statusCode = statusCode,
            isSuccess = isSuccess,
            requestSize = requestBodySize,
            responseSize = responseBodySize,
            duration = duration,
            timestamp = System.currentTimeMillis(),
            error = exception?.message,
            headers = request.headers.toMultimap()
        )

        storeNetworkEvent(networkEvent, sdkState)
    }

    /**
     * Store network event.
     */
    private fun storeNetworkEvent(event: NetworkEvent, sdkState: SdkState) {
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            try {
                sdkState.storageManager.storeEvent(
                    eventType = "network",
                    eventData = event,
                    sessionId = event.sessionId,
                    userId = event.userId
                )
                Logger.d("Network event stored: ${event.method} ${event.url} - ${event.statusCode} (${event.duration}ms)")
            } catch (e: Exception) {
                Logger.e("Error storing network event", e)
            }
        }
    }
}

data class NetworkEvent(
    val sessionId: String,
    val userId: String?,
    val url: String,
    val method: String,
    val statusCode: Int,
    val isSuccess: Boolean,
    val requestSize: Long,
    val responseSize: Long,
    val duration: Long,
    val timestamp: Long,
    val error: String?,
    val headers: Map<String, List<String>>
)

File: ios-sdk/Sources/MxLSdk/Network/NetworkMonitor.swift
import Foundation

/**
 * Network monitor for tracking network requests and responses.
 */
internal class NetworkMonitor: NSObject, URLSessionTaskDelegate {
    private let configuration: SdkConfiguration
    private var taskMetrics: [Int: URLSessionTaskMetrics] = [:]
    
    init(configuration: SdkConfiguration) {
        self.configuration = configuration
        super.init()
    }
    
    /**
     * Create URLSession with monitoring.
     */
    static func createSession(configuration: SdkConfiguration) -> URLSession {
        let sessionConfig = URLSessionConfiguration.default
        let monitor = NetworkMonitor(configuration: configuration)
        
        return URLSession(
            configuration: sessionConfig,
            delegate: monitor,
            delegateQueue: nil
        )
    }
    
    // MARK: - URLSessionTaskDelegate
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        guard configuration.enableNetworkMonitoring else { return }
        
        let sdkState = MxLSdk.getSdkState()
        guard let sdkState = sdkState else { return }
        
        guard let request = task.originalRequest,
              let url = request.url else {
            return
        }
        
        let metrics = taskMetrics[task.taskIdentifier]
        let response = task.response as? HTTPURLResponse
        
        let startTime = metrics?.taskInterval.start ?? Date()
        let endTime = metrics?.taskInterval.end ?? Date()
        let duration = (endTime.timeIntervalSince(startTime) * 1000) // Convert to milliseconds
        
        let requestSize = request.httpBody?.count ?? 0
        var responseSize = 0
        
        if let response = response,
           let contentLength = response.value(forHTTPHeaderField: "Content-Length"),
           let length = Int(contentLength) {
            responseSize = length
        }
        
        let networkEvent = NetworkEvent(
            sessionId: sdkState.sessionManager.getSessionId() ?? "unknown",
            userId: sdkState.sessionManager.getUserId(),
            url: url.absoluteString,
            method: request.httpMethod ?? "GET",
            statusCode: response?.statusCode ?? -1,
            isSuccess: (200...299).contains(response?.statusCode ?? 0),
            requestSize: Int64(requestSize),
            responseSize: Int64(responseSize),
            duration: Int64(duration),
            timestamp: Date().timeIntervalSince1970 * 1000,
            error: error?.localizedDescription,
            headers: request.allHTTPHeaderFields ?? [:]
        )
        
        storeNetworkEvent(networkEvent, sdkState: sdkState)
        
        // Clean up
        taskMetrics.removeValue(forKey: task.taskIdentifier)
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didFinishCollecting metrics: URLSessionTaskMetrics) {
        taskMetrics[task.taskIdentifier] = metrics
    }
    
    /**
     * Store network event.
     */
    private func storeNetworkEvent(_ event: NetworkEvent, sdkState: SdkState) {
        // TODO: Store in database and queue for upload
        Logger.shared.debug("Network event: \(event.method) \(event.url) - \(event.statusCode) (\(event.duration)ms)")
    }
}

struct NetworkEvent {
    let sessionId: String
    let userId: String?
    let url: String
    let method: String
    let statusCode: Int
    let isSuccess: Bool
    let requestSize: Int64
    let responseSize: Int64
    let duration: Int64
    let timestamp: TimeInterval
    let error: String?
    let headers: [String: String]
}

